File: src/content/docs/topics/cs/design-patterns/singleton.py
from __future__ import annotations


# Approach 1: __new__ override
class AppConfig:
    _instance: AppConfig | None = None

    def __new__(cls) -> AppConfig:
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance._settings: dict[str, str] = {}
        return cls._instance

    def get(self, key: str) -> str | None:
        return self._settings.get(key)

    def set(self, key: str, value: str) -> None:
        self._settings[key] = value


config1 = AppConfig()
config2 = AppConfig()
print(config1 is config2)  # True

config1.set('env', 'production')
print(config2.get('env'))  # production


# Approach 2: module-level instance (most Pythonic)
class _Config:
    def __init__(self) -> None:
        self._settings: dict[str, str] = {}

    def get(self, key: str) -> str | None:
        return self._settings.get(key)

    def set(self, key: str, value: str) -> None:
        self._settings[key] = value


config = _Config()
# Other modules do: from config import config
# The underscore on _Config signals that only the instance is public.

File: src/content/docs/topics/cs/design-patterns/singleton.ts
// Approach 1: classic GoF Singleton
class AppConfig {
  private static instance: AppConfig | null = null;
  private settings = new Map<string, string>();

  private constructor() {
    // Load from environment at construction time.
    this.settings.set('env', process.env.NODE_ENV ?? 'development');
    this.settings.set('port', process.env.PORT ?? '3000');
  }

  static getInstance(): AppConfig {
    if (!AppConfig.instance) {
      AppConfig.instance = new AppConfig();
    }
    return AppConfig.instance;
  }

  get(key: string): string | undefined {
    return this.settings.get(key);
  }

  set(key: string, value: string): void {
    this.settings.set(key, value);
  }
}

const config1 = AppConfig.getInstance();
const config2 = AppConfig.getInstance();
console.log(config1 === config2); // true

config1.set('feature_x', 'enabled');
console.log(config2.get('feature_x')); // enabled

// Approach 2: module-level instance (idiomatic TypeScript/Node.js)
class Config {
  private settings = new Map<string, string>();

  get(key: string): string | undefined {
    return this.settings.get(key);
  }

  set(key: string, value: string): void {
    this.settings.set(key, value);
  }
}

export const config = new Config();
// Node.js loads each module once and caches it.
// Every import of this module receives the same `config` object.
